src/sciline/graph.py

```diff
@@ -0,0 +1,70 @@
+# SPDX-License-Identifier: BSD-3-Clause
+# Copyright (c) 2023 Scipp contributors (https://github.com/scipp)
+from typing import (
+    Any,
+    Callable,
+    Dict,
+    List,
+    NewType,
+    Tuple,
+    TypeVar,
+    get_args,
+    get_origin,
+)
+
+from graphviz import Digraph
+
+from .pipeline import Graph
+
+
+def to_graphviz(graph: Graph) -> Digraph:
+    """
+    Convert output of :py:class:`sciline.Pipeline.get_graph` to a graphviz graph.
+
+    Parameters
+    ----------
+    graph:
+        Output of :py:class:`sciline.Pipeline.get_graph`.
+    """
+    dot = Digraph(strict=True)
+    for p, (args, ret) in _format_graph(graph).items():
+        dot.node(p, p, shape='ellipse')
+        for arg in args:
+            dot.node(arg, arg, shape='rectangle')
+            dot.edge(arg, p)
+        dot.node(ret, ret, shape='rectangle')
+        dot.edge(p, ret)
+    return dot
+
+
+def _format_graph(graph: Graph) -> Dict[str, Tuple[List[str], str]]:
+    return {
+        _format_provider(provider, bound=bound): (
+            [_format_type(a) for a in args.values()],
+            _format_type(ret),
+        )
+        for ret, (provider, bound, args) in graph.items()
+    }
```

Comment:
Not hugely relevant since graphs are small, but could yield tuples here instead of building a temporary dict.